### PR TITLE
Gb/multi file year

### PR DESCRIPTION
**Pull request by grantbuster**

### CODE REVIEW

File: rex/multi_year_resource.py [hunk 1]
```diff
@@ -123,15 +127,54 @@ def time_index(self):
...
+        h5_map : pd.DataFrame
+            DataFrame mapping file paths to open resource instances and
+            datasets per file (columns: fp, h5, and dsets)
+        """
+
+        h5_map = pd.DataFrame({'fp': file_paths, 'year': years, 'h5': None})
+        h5_map = h5_map.sort_values('year').reset_index(drop=True)
+
+        if len(h5_map['year'].unique()) < len(h5_map):
+            del res_cls_kwargs['hsds']  # no multi file res on hsds
```
Comment (ppinchuk) [inline rex/multi_year_resource.py:162]: Will there never be multi-file res on hsds? Is it fundamentally disallowed or is this just a reflection of the current state?

File: rex/multi_year_resource.py [hunk 2]
```diff
@@ -153,30 +196,24 @@ def _get_years(file_paths, years):
...
+            raise RuntimeError('Requested years "{}" not all found in '
+                               'file years "{}"'.format(years, fp_years))
+        else:
+            filtered_fps = []
+            years = sorted([int(y) for y in years])
+            for target_year in years:
+                for fp_year, fp in zip(fp_years, file_paths):
+                    if int(target_year) == int(fp_year):
+                        filtered_fps.append(fp)
+                        break
```
Comment (ppinchuk) [inline rex/multi_year_resource.py:215]: So does this mean that you can have multiple years in one file but not one year over multiple files?

File: rex/multi_year_resource.py [hunk 3]
```diff
@@ -279,14 +318,18 @@ def close(self):
...
+        for f in self._h5_map['h5']:
             f.close()
 
 
 class MultiYearResource(MultiTimeResource):
     """
     Class to handle multiple years of resource data stored accross multiple
-    .h5 files
+    .h5 files. This also works if each year is split into multiple files each
+    containing different datasets (e.g. for Sup3rCC and hi-res WTK+NSRDB).
```
Comment (ppinchuk) [inline rex/multi_year_resource.py:329]: Sorry for being dense here. This only allows a single year to be split across multiple files if the datasets are different, right? We still don't allow to split within a year?

So for example, we *do* allow this configuration:
 - file 1: 2012 data for windspeed 
 - file 2; 2012 data for wind direction

and we *don't* allow this configuration:
 - file 1: Jan - June 2012 data for wind speed and wind direction
 - file 2: July - Dec 2012 data for wind speed and wind direction

Is that right? In either case, we should consider making this docstring even more clear with this regard.

File: tests/test_multi_year_resource.py [hunk 4]
```diff
@@ -131,6 +133,8 @@ def check_time_index(res_cls):
 
     test = res_cls.time_index
 
+    print(test)
```
Comment (ppinchuk) [inline tests/test_multi_year_resource.py:136]: Same as above

File: rex/multi_time_resource.py [hunk 5]
```diff
@@ -299,12 +306,19 @@ def _map_file_instances(file_paths, res_cls=Resource, **res_cls_kwargs):
...
         """
-        h5_map = {}
-        for f_path in file_paths:
-            h5_map[f_path] = res_cls(f_path, **res_cls_kwargs)
+        h5_map = pd.DataFrame({'fp': file_paths, 'h5': None,
+                               't0': None, 't1': None})
+        for i, f_path in enumerate(file_paths):
+            h5_map.at[i, 'h5'] = res_cls(f_path, **res_cls_kwargs)
+            h5_map.at[i, 't0'] = h5_map.at[i, 'h5'].time_index.values[0]
+            h5_map.at[i, 't1'] = h5_map.at[i, 'h5'].time_index.values[1]
```
Comment (ppinchuk) [inline rex/multi_time_resource.py:318]: What is the point of `t1` here? I can't seem to find it used anywhere else in the code

File: rex/multi_file_resource.py [hunk 6]
```diff
@@ -614,3 +615,12 @@ def preload_SAM(cls, h5_source, sites, hub_heights, unscale=True,
...
 
         return SAM_res
+
+
+MULTI_FILE_CLASS_MAP = {Resource: MultiFileResource,
+                        NSRDB: MultiFileNSRDB,
+                        SolarResource: MultiFileResource,
+                        WindResource: MultiFileWTK,
+                        WaveResource: MultiFileResource,
+                        GeothermalResource: MultiFileResource,
```
Comment (ppinchuk) [inline rex/multi_file_resource.py:625]: Thanks for including other technologies besides wind and solar here :)

**Review by ppinchuk — COMMENTED** [review]
Using a pandas DataFrame to do a multi-level mapping is brilliant... I've never though about it. I always used to juggle around two or more dictionaries, but I will definitely be using this trick moving forward. 

Nice work overall! I just have a few questions below

File: tests/test_multi_time_resource.py [hunk 7]
```diff
@@ -141,6 +141,8 @@ def check_time_index(res_cls):
 
     test = res_cls.time_index
 
+    print(test)
```
Comment (ppinchuk) [inline tests/test_multi_time_resource.py:144]: Delete or convert to logging?

File: rex/multi_year_resource.py (hunk 1 above)
Comment (grantbuster) [reply rex/multi_year_resource.py:162]: loading data into HSDS is basically an exercise in indexing (as far as i can tell), so all files for a given domain get bundled into a single HSDS "filepath".

File: rex/multi_year_resource.py (hunk 2 above)
Comment (grantbuster) [reply rex/multi_year_resource.py:215]: For the multi-year handlers, yes. I'm not sure why it was originally designed this way but the multi-time handlers do multiple files within a year.

File: rex/multi_year_resource.py (hunk 3 above)
Comment (grantbuster) [reply rex/multi_year_resource.py:329]: I'm not sure why it was originally designed this way but the multi-time handlers do multiple files within a year. Will add to the docstring.

File: rex/multi_time_resource.py (hunk 5 above)
Comment (grantbuster) [reply rex/multi_time_resource.py:318]: it's really information/debugging purposes haha. This h5map is a great way to store meta data in-situ and the t0/t1 are both useful in this regard.

File: rex/multi_year_resource.py (hunk 2 above)
Comment (ppinchuk) [reply rex/multi_year_resource.py:218]: Ah okay, gotcha. We allow so many configurations these days that it's becoming harder to keep track! I will go familiarize myself more with the multi-time handlers. Thanks!

Comment (grantbuster) [reply rex/multi_year_resource.py:218]: Yeah i'm not sure if they're ever used in our work... I think they were developed for some misc wind data michael was working with. Fine to keep em for now but at some point we could probably merge the multi-time and multi-year handlers (they seem like they should fit into a common framework).

Comment (ppinchuk) [reply rex/multi_year_resource.py:218]: Right, the blurry distinction was probably why I was getting all confused. In any case, I think what you have for the time-being looks great!

**Pull request merged**